File: common/core/src/main/java/com/gadarts/te/common/utils/GeneralUtils.java
package com.gadarts.te.common.utils;

import com.badlogic.gdx.utils.Disposable;

import java.lang.reflect.Field;
import java.util.Arrays;

public final class GeneralUtils {
    public static <T> void disposeObject(T instance, Class<T> clazz) {
        Field[] fields = clazz.getDeclaredFields();
        Arrays.stream(fields).forEach(field -> {
            if (Disposable.class.isAssignableFrom(field.getType())) {
                field.setAccessible(true);
                try {
                    Object fieldValue = field.get(instance);
                    if (fieldValue instanceof Disposable) {
                        ((Disposable) fieldValue).dispose();
                    }
                } catch (IllegalAccessException e) {
                    throw new RuntimeException(e);
                }
            }
        });
    }

}

File: editor/core/src/main/kotlin/com/gadarts/te/GeneralUtils.kt
package com.gadarts.te

import com.badlogic.gdx.utils.Disposable
import java.lang.reflect.Field
import kotlin.reflect.KClass

object GeneralUtils {
    fun <T : Any> disposeObject(instance: T, kClass: KClass<T>) {
        val fields: Array<Field> = kClass.java.declaredFields
        for (field in fields) {
            if (Disposable::class.java.isAssignableFrom(field.type)) {
                field.isAccessible = true
                val fieldValue = field.get(instance)
                if (fieldValue is Disposable) {
                    fieldValue.dispose()
                }
            }
        }
    }

}
